// @kentcdodds/react-workshop-app/dist/styles/colors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.orange = exports.danger = exports.green = exports.yellow = exports.indigoLighten80 = exports.indigoDarken10 = exports.indigo = exports.gray80 = exports.gray20 = exports.gray10 = exports.gray = exports.text = exports.base = void 0;
const base = 'white';
exports.base = base;
const text = '#434449';
exports.text = text;
const gray = '#f1f2f7';
exports.gray = gray;
const gray10 = '#f1f1f4';
exports.gray10 = gray10;
const gray20 = '#e4e5e9';
exports.gray20 = gray20;
const gray80 = '#6f7077';
exports.gray80 = gray80;
const indigo = '#3f51b5';
exports.indigo = indigo;
const indigoDarken10 = '#364495';
exports.indigoDarken10 = indigoDarken10;
const indigoLighten80 = '#b7c1f8';
exports.indigoLighten80 = indigoLighten80;
const yellow = '#ffc107';
exports.yellow = yellow;
const green = '#4caf50';
exports.green = green;
const danger = '#ef5350';
exports.danger = danger;
const orange = 'orange';
exports.orange = orange;
